### DISCUSSION TITLE
SDR→HDR via RTX HDR (upmapping / inverse tone mapping)

### DISCUSSION THREAD

**Discussion by alexandergunnarson** [libvips/libvips#3907] [Ideas]
First off, let me say that I'm grateful for your work on Vips. We use it in production at my company and are very happy with it. An extremely well-thought-out and versatile tool!

So, this may not be quite the right library for this feature request, but throwing it out there anyway to at least spark a discussion.

I'm looking to upmap ("upscale") images from SDR to HDR. This would potentially be quite a useful feature for a number of people, as there is a whole subfield of AI research devoted to this (AI-ITM / AI inverse tone mapping).

ML models are of course out of scope of this repo, but NVIDIA [released RTX HDR earlier this year](https://blogs.nvidia.com/blog/rtx-video-hdr-remix-studio-driver/), which can automatically upmap video from SDR to HDR on Windows machines with RTX cards. (There may of course be other system requirements here.) It apparently performs quite well, and operates at the driver level. There's another repo that [appears to reach into this driver in C++](https://github.com/Aleksoid1978/VideoRenderer/commit/b9f82f315b30c5e30a0fd47f2ef7b1e0d65ad3f2), presumably in the course of managing a video stream.

The idea here to achieve the SDR→HDR upmapping — on supported systems — is to 1) render an image to a single-frame video with RTX HDR applied and 2) output the resulting frame for use downstream within Vips.

[This article by Dolby from 2021](https://professionalsupport.dolby.com/s/article/Inverse-tone-mapping-upscaling-SDR-content-to-HDR?language=en_US) appears to be a good authoritative resource around SDR→HDR.

For completeness, I'll mention that there are three ways I'm aware of to do SDR→HDR:
- Doing mathematical conversion via "widening" (the Dolby article mentions [this source](https://www.movielabs.com/ngvideo/MovieLabs_Mapping_BT.709_to_HDR10_v1.0.pdf) as being relevant; it's simple but far inferior to "true" upmapping)
- Using an ML model (e.g. https://github.com/AndreGuo/ITMLUT)
- NVIDIA RTX HDR

Thanks for your time!

**Comment by jcupitt** [comment]
Hi @alexandergunnarson,

Yes, this has been done a few times -- xbox has something similar, though much simpler. I think we'd need a cross-platform, free implementation before we could build it into the core library.

**Comment by alexandergunnarson** [comment]
Makes sense! You mention "this has been done a few times" — curious what other implementations are out there that you're aware of. Basically I just want an API of some sort (preferably CLI, but HTTP API is fine too) that takes an SDR image and upmaps to HDR. From my search, this level of plug-and-play doesn't yet exist.

**Reply by reconlabs-sergio** [reply]
@alexandergunnarson , did you find any tool for this? I'm in the same quest...

**Reply by alexandergunnarson** [reply]
[TEKNO3D HDR-X](https://tekno3d.com/)!
